# Практическая работа 3. Обработка данных из файла

#Выполнил студент ПИН-Б-З-22-1 Гадиян Сергей Гариевич
## Индивидуальное задание находится внизу

## Задание

в файле data_v1.txt содержится последовательность символов, содержащая буквы латинского алфавита и десятичные цифры. Необходимо решить следующие задачи:
1. Найдите максимальное четное число в строке (под числом понимается последовательность цифр, ограниченная буквами).
2. Найдите буквенную подпоследовательность максимальной длины (то есть подпоследовательность не содержащую цифр). В ответе укажите ее длину.
3. Найдите символ в файле, который встречается чаще всего. Если таких символов несколько, то выведите все эти символы, в конце укажите одно число – сколько раз встречаются эти символы.


## Анализ задания, получение данных

Файл data_v1.txt расположен репозитории. Выполним копирование файла в локальное хранилище Google Colab, используя следующую команду:

In [4]:
!wget https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v1.txt

--2025-06-10 20:49:22--  https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v1.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 100002 (98K) [text/plain]
Saving to: ‘data_v1.txt’

data_v1.txt         100%[===================>]  97.66K  --.-KB/s    in 0.02s   

2025-06-10 20:49:23 (3.92 MB/s) - ‘data_v1.txt’ saved [100002/100002]



Реализуем чтение всего файла в строку с использованием метоа `open()` и `read()`:

In [ ]:
f=open('data_v1.txt')
s=f.read().strip()
print(len(s))

100000


## Обработка символьной последовательности

Строки, подобно списка, являются коллекциями. Только в качестве элементов последовательности выступают символы. Для строк также используются функции нахождения длины `len()`, операция индексации `[]`

Например, можно посимвольно обходить строку, используя цикл `for` и индекс строки:

In [ ]:
# Вывод первых 10 символов строки с одновременным анализом буква это или цифра
for i in range(10):
  c = s[i]
  if c.isdigit():
    t = 'цифра'
  else:
    t='буква'
  print(c, t)

Z буква
S буква
t буква
M буква
P буква
i буква
7 цифра
c буква
s буква
i буква


## Решение задачи 1

Будем посимвольно обходимть строку и накапливать символы-цифры в строке `target`. Если цифровая последовательность прервется, то конвертируем `target` в число `int` и проверяем является ли число четным и не превысило ли оно максимум

In [ ]:
target=''
maxx=0
for c in s:
  if c.isdigit():
    target+=c
  else:
    if target!='':
      x = int(target)
      maxx=max(x, maxx)
      target=''

print('Максимальное четное число = %d' % maxx)


Максимальное четное число = 838478


## Решение задачи 2

Решение данной задачи осуществляется аналогично предыдущей. Можно даже вывести самуюдлинную буквенную строку

In [ ]:
target=''
maxx=0
maxs=''
for c in s:
  if c.isalpha():
    target+=c
  else:
    if len(target)>maxx:
      maxx=len(target)
      maxs=target
      target=''

print('Максимальная длина буквенной строки = %d' % maxx)
print('Сама строка:', maxs)

Максимальная длина буквенной строки = 1025
Сама строка: rpfunCMcMCfmnlNBbiCkCOPQsRYZmLjneFWZJjHqUxiTFGtMiKUTvaUkXBQnvGgpnDmUtFNyXZwWIdyqBWzLEqsTqoXCOkPaFQTuVxYeTftNLcEEjnRUqqAfeOmBHdcIDfyBpmvYjIivcZXMaflpfECMZKprJUSPIJtoolmBwWFmTiXtLsUgZrirYdtPUGEhsmAgGwqxoTcNlbtFmVApQAkcRiRFbvQxKwpdEVZhgUAymnZnRnJVmnDIvHmxtJkiICukqvSsMRrAHDBsAFDBWieGPhTcXNdetYzIdITVgvyiPqhPTKtixxggpYfCThpmcEqclpTTjnnHAXbpWOaOPlGcabgWvgiDDOhSWdTviUlNvTaoEVhoBfgrJkoJxtyDWXjVYYRokqkKPwRdkPkQYaVIoHCAfYPEyHgBsbUkYQyKNfgyWNJJsmtAoTPGrHKwUyJDvCEqgeRLTFdpqxaZFFGJrmkQuqbyadKxgWFsznrBtZyuubYwcmDAMqCqWXSUomuCMcFWWoxAWMJfdcYlDGMDSSCnYCUOVvgENGKZiTXKuLaLtMHTPjKmWwjwhfaePXiOPwZdvnAlPruuGwKhbRevjOHPVWIShguJJgKLaqUMJLwokvxJzYpvfOvfzRDHGQbnrWxESwCdsUzxyFWEYqRXpihePHiHXbMXBYyvIcUlaudLjXIPdXdMHrExnGgHeanzgtDSgSGOkxupzNQitVYWFXSNOpsdKMAveZQieNAqCMEXZzBdemldlsWCpiIMuQomlFEbZKcBwxAbNlFGPejoMQxgPHzXiskhadywliHoMmZlokYoauVlxmuHrqhwJHYpgzfUAqwPBAQPJNKlKhhYCRrLocFcIEorvZnBPIjlLVtMzafAoNJRufesTkbJcvviKEOSowgawXzWsFeUYgGsrdXZHbRGVQGvSczyAvdAeLjS

## Решение задачи 3

Для решения данной задачи будем использовать коллекцию словарь `dict()`. Данная коллекция содержит элемкнты в виде пар `ключ : значение`. Доступ к элементам осуществляется по ключу с использованием операции ндексации.

Заполним словарь d, который после прохода по строке должен принять вид:
d = {'A' : 10, 'B' : 127, '3' : 217, ... }
то есть в качестве ключа будем использовать символ, а вкачестве значения - количество таких символов в строке:

In [ ]:
# Создаем пустой словарь
d={}
# Проходим по строке и заполняем словарь
for c in s:
  d[c] = d.get(c, 0) + 1

# Выводим получившийся словарь:
print(d)

{'Z': 1623, 'S': 1613, 't': 1711, 'M': 1618, 'P': 1670, 'i': 1542, '7': 1643, 'c': 1562, 's': 1606, 'T': 1699, 'b': 1555, 'r': 1581, 'm': 1686, 'E': 1662, 'h': 1596, '8': 1654, 'f': 1609, 'B': 1647, 'K': 1616, 'a': 1586, 'F': 1623, 'O': 1580, 'w': 1582, 'o': 1609, '5': 1631, '1': 1580, 'W': 1595, '6': 1597, '0': 1651, 'I': 1604, 'u': 1606, 'e': 1566, 'p': 1644, 'D': 1644, 'A': 1638, 'Q': 1544, 'J': 1638, 'U': 1629, 'R': 1590, 'C': 1570, 'G': 1597, 'k': 1618, 'j': 1603, 'N': 1636, 'l': 1583, '2': 1600, 'q': 1614, 'X': 1619, 'H': 1565, '3': 1587, 'L': 1635, 'Y': 1647, '9': 1611, 'V': 1601, 'x': 1654, 'y': 1593, 'g': 1603, '4': 1599, 'v': 1655, 'd': 1599, 'z': 1615, 'n': 1566}


В получившемся словаре необходимо найти самый частотный символ. Для этого не будем обходить словарь, а просто воспользуемся встроенной функцией `max()`, которую применим к коллекции `values()` словаря

In [ ]:
max_val = max(d.values())
print(max_val)

1711


Выполним обход словаря и напечатаем ответ к данной задаче:

In [ ]:
for k in d:
  if d[k]==max_val:
    print(k, end=' ')
print(max_val)

t 1711


Для данного файла только один символ `t` встречается в файле максимальное `1711` количество раз

В данной работе были использованы операторы `if` (условный оператор), `for ... in range(...)` (оператор цикла с обходом диапазона). Применялись функции `print()`, `len()`, `max()`. В работе использовался тип данных `dict()`

 # Индивидуальное задание

##Вариант 4
В файле data_v1.txt найдите количество чисел кратных 3 но не
кратных 9 (под числом понимается непрерывная
последовательность цифр). В ответе укажите количество таких
чисел и максимальное из них.

In [5]:
# открываем файл для чтения и считываем его содержимое
with open('data_v1.txt', 'r') as f:
    s = f.read().strip()  # удаляем пробелы в начале и конце строки
# список для хранения извлеченных чисел
numbers = []
# временная строка для сбора цифр текущего числа
target = ''
# проходим по каждому символу в строке
for c in s:
    if c.isdigit():
        # если символ - цифра, добавляем его к target
        target += c
    else:
        # если символ не цифра и target не пуст (было собрано число)
        if target:
            # преобразуем target в целое число и добавляем в список numbers
            numbers.append(int(target))
            # очищаем target для сбора следующего числа
            target = ''
# после цикла проверяем, осталась ли незавершенная последовательность цифр в target
if target:
    # добавляем последнее число в numbers
    numbers.append(int(target))
# список для хранения чисел, удовлетворяющих условию (кратны 3, но не кратны 9)
valid_numbers = []
# проходим по каждому числу в списке numbers
for x in numbers:
    # проверяем условие: число делится на 3, но не делится на 9
    if x % 3 == 0 and x % 9 != 0:
        # добавляем число в список valid_numbers
        valid_numbers.append(x)
# подсчитываем количество чисел, удовлетворяющих условию
count = len(valid_numbers)
# находим максимальное число среди valid_numbers, если список не пуст
max_num = max(valid_numbers) if valid_numbers else None
# выводим результаты
print(f"Количество чисел: {count}")
print(f"Максимальное число: {max_num}")

Количество чисел: 2766
Максимальное число: 73758


##Ответы на контрольные вопросы

**1. Что общего у строки и списка в Python? <br>**

Оба являются последовательностями (iterable):
Можно использовать циклы for для обхода элементов.

Индексация: Доступ к элементам по индексу (например, s[0] для строки s и lst[0] для списка lst).

Срезы: Поддерживают срезы (например, s[1:5] и lst[1:5]).

Функция len():
Возвращает длину строки или списка.


 **2. Перечислите и охарактеризуйте функции класса str в Python.<br>**

str.isdigit():

Проверяет, состоит ли строка только из цифр.

str.isalpha():

Проверяет, состоит ли строка только из букв.

str.strip():

Удаляет пробелы в начале и конце строки.

str.split(sep=None):

Разделяет строку на список подстрок по разделителю sep.

str.join(iterable):

Объединяет элементы итерируемого объекта в строку, используя текущую строку как разделитель.

str.lower() / str.upper():

Преобразует строку к нижнему/верхнему регистру.

str.replace(old, new):

Заменяет все вхождения подстроки old на new.

**3. Охарактеризуйте тип данных dict в Python. <br>**
Словарь (dict): Хранит данные в виде пар ключ: значение.

Ключи: Должны быть уникальными и неизменяемыми (например, строки, числа).

Значения: Могут быть любого типа данных.

Динамический: Можно добавлять, удалять и изменять элементы.

**4. Перечислите и опишите использование функций класса dict.<br>**
dict.get(key, default=None):
 - Возвращает значение по ключу key. Если ключа нет, возвращает default.

dict.keys():

- Возвращает объект dict_keys с ключами словаря.

dict.values():

- Возвращает объект dict_values со значениями словаря.

dict.items():

- Возвращает объект dict_items с парами ключ: значение.

dict.update(other_dict):

- Обновляет словарь элементами из other_dict.

dict.pop(key, default):

- Удаляет элемент по ключу и возвращает его значение. Если ключа нет, возвращает default.

**5. Опишите механизм чтения данных из текстового файла. <br>**
Открытие файла:<br>
`f = open("file.txt", "r")  # r — режим чтения` <br>

Чтение содержимого:

`f.read()` — читает весь файл как одну строку.

`f.readline()` — читает одну строку.

`f.readlines()` — читает все строки в список.

Закрытие файла:<Br>
`f.close()`

**6.Опишите назначение, сходства и различия функций чтения из текстового
файла: read( ), readline( ), readlines( ).<br>**
Сходства:

- Все методы читают данные из файла.

- Возвращают данные в виде строк.

Различия:

- read() читает весь файл за раз.

- readline() читает построчно.

- readlines() читает все строки в список.


**7. Как прочитать многострочный текстовый файл не используя явно методы
read( ), readline( ), readlines( )? <br>**
Можно использовать итерацию по самому файлу — он является итерируемым объектом:

```
with open("file.txt", "r") as f:
    for line in f:
        print(line, end='')  # end='' чтобы не добавлялся лишний перевод строки
```
